src/persist-client/src/internal/gc.rs

```diff
@@ -205,273 +209,224 @@ where
             step_start = now;
         };
-
         assert_eq!(req.shard_id, machine.shard_id());
-        // NB: Because these requests can be processed concurrently (and in
-        // arbitrary order), all of the logic below has to work even if we've
-        // already gc'd and truncated past new_seqno_since.
 
         let mut states = machine
             .applier
             .state_versions
-            .fetch_all_live_states::<T>(req.shard_id)
+            .fetch_all_live_states(req.shard_id)
             .await
-            .expect("gc should only be called on an initialized shard")
-            // TODO: Consider pulling the K, V, D params off of GC. If we do,
-            // then we should be able to delete TypedStateVersionsIter (and
-            // probably merge UntypedStateVersionsIter into StateVersionsIter).
-            .check_codecs::<K, V, D>()
-            .expect("shard codecs should not change");
+            .expect("state is initialized")
+            .check_ts_codec()
+            .expect("ts codec has not changed");
+        report_step_timing(&machine.applier.metrics.gc.steps.fetch_seconds);
+
+        machine
+            .applier
+            .shard_metrics
+            .gc_live_diffs
+            .set(u64::cast_from(states.len()));
 
         debug!(
-            "gc {} for {} got {} versions from scan",
-            req.shard_id,
+            "gc seqno_since: ({}) got {} versions from scan",
             req.new_seqno_since,
             states.len()
         );
-        report_step_timing(&machine.applier.metrics.gc.steps.fetch_seconds);
 
-        let earliest_live_seqno = match states.peek_seqno() {
-            Some(x) => x,
-            None => panic!(
-                "initialized shard {} should have at least one live state",
-                req.shard_id
-            ),
-        };
-        // Fast-path: Someone already GC'd past `req.new_seqno_since`, don't
-        // bother running any of the below logic.
-        //
-        // Also a fix for #14580.
-        if earliest_live_seqno > req.new_seqno_since {
-            machine.applier.metrics.gc.noop.inc();
-            debug!(
-                "gc {} early returning, already GC'd past {}",
-                req.shard_id, req.new_seqno_since,
+        Self::incrementally_gc(
+            &req.shard_id,
+            &mut states,
+            req.new_seqno_since,
+            machine,
+            &mut report_step_timing,
+        )
+        .await;
+
+        // WIP: do we want to remove rollups incrementally?
+        let removable_rollups = machine.applier.removable_rollups();
+        let (removed_rollups, maintenance) = machine.remove_rollups(&removable_rollups).await;
+        debug!("removed rollups: {:?}", removed_rollups);
+
+        // WIP: do we want to calculate the seqno_parts_held metric?
+        while let Some(_) = states.next(|_| {}) {}
+
+        maintenance
+    }
```

Comment:
make sure to put report_step_timing calls through the very end to make sure all the work is covered